ath79: generic: rework ar9342_ubnt_xw dtsi, and add support for Ubiquiti LiteBeam M5 (XW), Ubiquiti AirGrid M5 HP (XW), Ubiquiti PowerBeam M5 300 (XW)

a-gave opened the pull request:
These devices are almost identical and both can work with the firmware image of ubnt_nanostation-loco-xw
I wonder if there is some unexplicited habit of not adding more copies of the same device if at least an already present firmware image can run on it. For this reason this pr could be considered more a merge proposal than a merge request.

I'm trying to add them only because: 
1- the dish is different, and if one install the firmware of loco-xw it works but there is no mention anywhere in the device of what hardware it is exactly. In particular they have both around 15° of horizontal azimuth [0] [1] compared with the 45° of the ubnt_nanostation-loco-xw
2- they both have the minimum 8MB of flash and 64MB of ram, so they still could be used with latest openwrt 23, but on the wiki they are listed as stopped at 19.07.10
3- using them with the firmware image of another device implies more memory work than simply selecting the device on the openwrt firmware selector


[0] https://dl.ubnt.com/datasheets/airgridm/airGrid_HP.pdf
[1] https://dl.ubnt.com/datasheets/powerbeam/PowerBeam_DS.pdf


---

As a (sorry but long) side note:
On these devices, with openwrt-23.05.x, I'm ***not*** experiencing the 'configuration loss after reboot' found here https://forum.openwrt.org/t/powerbeam-m5-xw-configuration-loss-after-reboot/141925 and partly here https://github.com/openwrt/openwrt/issues/12882 (but this latter mention mostly AC devices)
But I saw it on another almost identical device (an ubnt_litebeam-m5-xw) that wasn't running this bootloader version "U-Boot 1.1.4-s776 (Nov 27 2013 - 15:58:45)" correspondent to the airos firmware XW.v5.5.10-u2.28005.150723.1358.bin (or XW.v5.5.10.24238.141001.1641.bin) or previous. In this case the issue was fixed downgrading the bootloader. 

There were no problems in downgrading the ubnt_airgrid-m5-hp-xw or ubnt_powerbeam-m5-300-xw to XW.v5.5.10-u2
The ubnt_litebeam-m5-xw (ar9342 v2) instead wasn't accepting the downgrade from XW.v5.6.15 to XW.v5.5.10-u2 unless via serial typing in the bootloader `urescue -f -e` but this implied loosing the waterproof.
So I managed to downgrade these devices doing a backup of /dev/mtd0 from a device that I opened and flashing then the bootloader via the `mtd write` command on openwrt-19.07.10 (because in openwrt-23.05.x the mtd0 is defined in the dts as readonly and i guess it would require a different build that unlock it)

```
opkg update
opkg install kmod-mtd-rw
insmod mtd-rw.ko i_want_a_brick=1
mtd -r write /tmp/uboot_litebeam.bin /dev/mtd0
```

As scary as it is the `i_want_a_brick` I would like to document this procedure in the wiki if anyone gives me positive feedback.
I've tested it downgrading from these versions that had the same bootloader `U-Boot 1.1.4-s958 (Jun 10 2015 - 10:56:20)` that correspond to airos XW.v5.6.1 and XW.v5.6.15 and XW.v6.0.4-cs

I didn't tested this procedure so far with devices that were flashed without a proper downgrade and that runs openwrt 19.07.10 with these bootloader versions:
- `U-Boot 1.1.4-s1039 (May 24 2017 - 15:58:18)` that correspond to airos XW.v6.1.7
- `U-Boot 1.1.4-gb02ecccc (Feb 14 2020 - 10:53:11)` that correspond to airos XW.v6.3.2.

neheb commented on target/linux/ath79/dts/ar9342_ubnt_airgrid-m5-hp-xw.dts:
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "ar9342_ubnt_xw.dtsi"
```
Can mtd-cal-data be replaced with nvmem here?

a-gave commented:
> If the devices are identical (firmware-wise), why not add support via DEVICE_ALTx_MODEL in the Makefile?

This is nice, I've tested it and it make it easier to find the firmware image on the selector.
However in this case it remains difficult later to distinguish them, since they are airmax devices with different capabilities:
- nanostation-loco-xw has 45° of horizontal azimuth and up to 15km of distance
- airgrid-m5-hp-xw has 15° of horizontal azimuth and up to 30km of distance
- powerbeam-m5-300-xw has 15° of horizontal azimuth and up to 3km of distance

I happened to go to an installation knowing that one device nearby had 45° horizontal azimuth and therefore there was no need to add a second one but you could use that one. I think that may be the case for other people as well.

raenye commented:
Not sure what you mean by distinguishing. If you bought device XYZ, you want to lookup XYZ in the list of downloads, right?

dnordenberg commented:
The boot loader downgrade was a nice catch for #12882 but it is not really a solution since I don't think there is a firmware with that old U-boot for AC devices. And saw open devices to serial flash unsupported firmware isn't really nice either :( But it is nice to see people finding workarounds :)

a-gave committed:
ath79: generic: rework ar9342_ubnt_xw dtsi, and add support for Ubiquiti LiteBeam M5 (XW), Ubiquiti AirGrid M5 HP (XW), Ubiquiti PowerBeam M5 300 (XW)

Add support for Ubiquiti LiteBeam M5 (XW).
The device was previously supported in ar71xx.
See commit: https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=d0988235dd277b9a832bbc4b2a100ac6e821f577

Add ALTX_MODEL for Ubiquiti AirGrid M5 HP (XW), Ubiquiti PowerBeam M5 300 (XW) in generic-ubnt.mk
This models are identical (firmware-wise) to the already supported Ubiquiti Nanostation Loco M (XW)

Add also Ubiquiti NanoBeam M5 to ALTX_MODEL of Ubiquiti Nanostation Loco M (XW) since it's another clone.

Tested on:
- Ubiquiti LiteBeam M5 (XW)
- Ubiquiti PowerBeam M5 (XW)

This also modify target/ath79/dts/ar9342_ubnt_xw.dtsi to use nvmem for calibration data
Checked that the caldata size in the eeprom partition are actually 0x440 on:
- Ubiquiti PowerBeam M5 (XW)
- Ubiquiti Nanostation M5 (XW)
- Ubiquiti LiteBeam M5 (XW)
- Ubiquiti AirGrid M5 HP (XW)

Signed-off-by: Samuele Longhi <agave@dracaena.it>

a-gave commented:
This should now replace https://github.com/openwrt/openwrt/pull/14104

a-gave commented:
> Not sure what you mean by distinguishing. If you bought device XYZ, you want to lookup XYZ in the list of downloads, right?
Sorry for the delay, I was meaning (for distinguish) that after installing the firmware all these devices state (`ubus call system board`) they are Ubiquiti Nanostation Loco M5 (XW) and not the exact model they are.

One question, I initially found that some of these devices have the 'product name' stored in the u-boot-env partition, in this way:
<details><summary>root@openwrt:~# strings /dev/mtd1</summary>
<pre>
=bootdelay=1
baudrate=115200
ethaddr=0x00:0xaa:0xbb:0xcc:0xdd:0xee
mtdids=nor0=ath-nor0
filesize=40000
fileaddr=81000000
ethact=eth0
mtdparts=mtdparts=ath-nor0:256k(u-boot),64k(u-boot-env),1024k(kernel),6528k(rootfs),256k(cfg),64k(EEPROM)
bootcmd=bootm 0x9f050000
stdin=serial
stdout=serial
stderr=serial
serverip=192.168.1.254
ipaddr=192.168.1.20
bootargs=console=tty0 root=31:03 rootfstype=squashfs init=/init
product=AG5-HP
</pre>
</details> 

However not all ubnt_xw devices has such abbreviation in the u-boot-env (maybe only newers, I can't check)
I've tested on more than 50 devices among XW and XM ones, produced between 2012 and 2019
On devices without this information in the u-boot-env I imagine that some other code is stored in the eeprom since the `airos XW firmware` is anyway able to tell the exact model.
At the moment I can't say how they do because I found mixed results.

Before continuing, I wonder if it is possible to include in the firmware of these devices a package (lets' call it `ubnt_detectboard` or similar) that do a model discovery using this association `acronym -> full_model_name` to overwrite the "model.name" property in `/etc/board.json`? (not the "model.id" so sysupgrade should works normally) 

I wonder also if this packages would be suitable for inclusion in the buildroot and which would be the appropriate place to include it? 
Since it would be related only to a subset of ath79/generic that are ubnt-xw devices;
Should it be defined like these ones below, found searching for 'define Package' inside <owrt_buildroot>/target?
```
grep -r "define Package" target/
target/linux/x86/base-files.mk:define Package/base-files/install-target
target/linux/armsr/base-files.mk:define Package/base-files/install-target
```

Thanks for your reply

a-gave commented:
> The boot loader downgrade was a nice catch for #12882 but it is not really a solution since I don't think there is a firmware with that old U-boot for AC devices. And saw open devices to serial flash unsupported firmware isn't really nice either :( But it is nice to see people finding workarounds :)

Thanks! The workaround should not be needed anymore thanks to https://github.com/openwrt/openwrt/pull/14237 :) the device opened had been bricked for a few years by mistake, peace! At least it's bootloader served to not open the other devices

hauke commented on target/linux/ath79/image/generic-ubnt.mk:
```diff
@@ -112,8 +112,17 @@ TARGET_DEVICES += ubnt_nanostation-ac-loco
+  DEVICE_ALT2_VENDOR := Ubiquiti
+  DEVICE_ALT2_MODEL := NanoBeam M5
+  DEVICE_ALT2_VARIANT := XW
+  SUPPORTED_DEVICES += loco-m-xw nanostation-m-xw
```
Please add the `ubnt_nanobeam-m5-xw` to the supported devices. If someone had the old image installed only a force update would work, if it is added to `SUPPORTED_DEVICES` the new image will install without problems.
